### ISSUE TITLE
Issues with option settings on Self Test page.

### ISSUE THREAD

**Issue by bakerhillpins** [jessaskey/mhavocpe#174]
Some stuff that doesn't operate the way I expected when I tried the Self Test Option settings.

1.  Can't select reset to DIP settings unless you made a change to one of the options. Any change, even back to what it was.
2.  For some reason Free Play/Coin settings don't allow you to change them away from the DIP settings. And once you've made a choice in Self Test it will not allow you to change even if you actually change the DIP setting. The only way to get it to move is to select the reset to DIP and then it will follow the new board setting.
3. It's not clear what Difficulty you play in when in Demo Mode?

**Comment by jessaskey** [comment]
1. and .2 - I will check this out on my game too just to make sure I understand. 
3. So DEMO mode is special in two ways... it is always Medium difficulty and it is always FREE play.  Since demo mode is an option in the difficult DIPs... you can't select DEMO AND a difficulty, there has to be a default for it. I would change it, but then the DIPS start having totally different functions and while I have eliminated stuff like 2-coin minimum to save ROM space, the DIPS are pretty much the same. I *do* need to update the docs to have the final DIP settings documented tho.

**Comment by jessaskey** [comment]
Okay #1 is fixed in the v0.80 ROMs...  as to number 2 above... when changing settings in MAME, things seem to work as normal.

1. Hit fire to move to a setting.
2. Rotate roller to change setting (this will turn on the 'Hit Shield to Confirm' message when the setting is different from NVRAM
3. Hit Shield
4. Hit Shield again to confirm.

Setting is saved. Leaving self test and coming back shows the new setting still.

**Comment by bakerhillpins** [comment]
For 2 it must have something to do with reading the HW then. Because it makes the change on the screen but when you exit and return it returns to the previous.  The odd part is if you change the physical DIP sw, and then go into self test it will show the previously set val. You can try to change it but it behaves the same: It allows the change, when you exit and return it's back to the previous value (which is now different than the DIP). You then have to select Reset to DIPs to get it to change and stick.

**Comment by jessaskey** [comment]
So it is definitely confusing having both the DIPs AND the ability to change settings in the Self Test. I did this because it basically sucked to have to pull your cabinet out from the wall, flip your dips etc... especially for something like going into DEMO mode.... so that was why I did this originally, plus I thought it was 'neat'. 

So, the DIPS are the 'default' or in this case the 'fallback' if the data in the NVRAM get corrupted for any reason. The settings are stored in NVRAM with a checksum. If that checksum is wrong, then it throws away all the settings and reloads what the DIPS are set too.

With what you wrote above, tho, it seems like the writing back to NVRAM of the current setting in Self Test is not working correctly. I will look at this on my PCB tonight and see if I can figure it out. Odd that it works in MAME tho, but might be a timing issue then on the actual 2804 NVRAM.
